classes/output/login.php

```diff
@@ -49,51 +53,84 @@ public static function output_login_options($wantsurl) {
             }
             $authplugin = get_auth_plugin($authname);
             $potentialidps = $authplugin->loginpage_idp_list($wantsurl);
-
             if (!empty($potentialidps)) {
                 foreach ($potentialidps as $idp) {
-                    $output .= \html_writer::start_tag('div', array('class' => 'idp-login container-fluid '. $authname));
-                    $name = '';
+                    $idpcontext = [
+                        'auth' => $authname,
+                        'name' => s($idp['name']),
+                        'url' => $idp['url']
+                    ];
                     if (!empty($idp['iconurl'])) {
-                        $name .= '<img src="' . s($idp['iconurl']) . '" width="24" height="24" class="mr-1"/>';
+                        $idpcontext['iconurl'] = s($idp['iconurl']);
                     }
-                    $name .= s($idp['name']);
-                    $attributes = ['class' => 'btn btn-secondary btn-block', 'title' => s($idp['name'])];
-                    $output .= \html_writer::link($idp['url'], $name, $attributes);
-
-                    $output .= \html_writer::end_tag('div');
-                    $idploginpath = $idp['url'];
+                    $idplist[] = (object) $idpcontext;
                     $count++;
+                    $idploginpath = $idp['url'];
                 }
             }
         }
+        $context->idplist = $idplist;
+
         // If only one IDP is available in authentication plugins then auto-redirect to it.
         $noredirect  = optional_param('noredirect', 0, PARAM_BOOL); // Don't redirect.
         if ($count === 1 && get_config('local_login', 'autoredirect')  && empty($noredirect)) {
             redirect($idploginpath);
         }
 
+        // If data exists in the cache, use it if it is still valid.
+        // We need to check this after calculating the IDP data to ensure its current.
+        $cache = cache::make('local_login', 'renderedlogin');
+        $cachehash = $cache->get('idphash');
+        $idphash = sha1(json_encode($idplist));
+        $cacheval = $cache->get('html');
+        if ($cachehash === $idphash && !empty($cacheval)) {
+            return $cacheval;
+        }
+
         if (!empty($config->showmanual)) {
+            $manual = new \stdClass();
```

Comment:
Nitpick - a comment here explaining this might be useful - e.g. why do you manually create a manual login ?